File: src/BOJ9663.java
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.util.Scanner;

//백준 - N-Queen
//브루트포스, 백트래킹

public class BOJ9663 {
	static int N;
	static int[] arr;
	static boolean[] visit;
	static int cnt = 0;
 
	public static void nQueen(int depth) {
		
		if(depth == N) {
			cnt++;
			return;	
		}
		
		for(int i=0;i<N;i++) {
			if(visit[i])
                continue;

			arr[depth] = i; //배열 인덱스는 열, 값은 행
			
			if(isOk(depth)) {
				visit[i] = true;
				nQueen(depth+1);
				visit[i] = false;
			}
		}
		
	}
	
	public static boolean isOk(int depth) {
		
		for(int i=0;i<depth;i++) {
			if(arr[i] == arr[depth])
				return false;
			
			if(Math.abs(i - depth) == Math.abs(arr[i] - arr[depth])) {
				return false;
			}
		}
		
		return true;
	}
	
	public static void main(String[] args) throws FileNotFoundException {
		System.setIn(new FileInputStream("D:\\eclipse\\workspace\\Study\\WebContent\\9663.txt"));
		Scanner in = new Scanner(System.in);
		
		N = in.nextInt();
		arr = new int[N]; //값은 행, 배열은 열
		visit = new boolean[N];
 
		nQueen(0);
		System.out.println(cnt);
 
	}
}

File: src/BOJ9663.py
#백준 - N-Queen
#브루트포스, 백트래킹

import sys

def check(x):
    for i in range(x):
        if row[x] == row[i] or abs(row[x] - row[i]) == x - i:
            return False
    return True


def dfs(x):
    global result
    if x == N:
        result += 1
        return

    for i in range(N):
        if visit[i]:
            continue

        row[x] = i
		
        if check(x):
            visit[i] = True
            dfs(x + 1)
            visit[i] = False

N = 8
row = [0] * N
visit = [False] * N
result = 0
dfs(0)
print(result)